### PR TITLE
Build: add commit message linting

### PR DESCRIPTION
**Pull request by jamesramsay**
Add commit message linting which checks:

- one commit per PR
- length of first line (72 chars)
- prefix of first line of commit (`PREFIX: short message`)

Proposed allowed commit prefixes (based on ESLint commit message style):

- `Fix` - for a bug fix. (PATCH)
- `New` - implemented a new feature. (MINOR)
- `Update` - for a backwards-compatible enhancement. (MINOR)
- `Breaking` - for a backwards-incompatible enhancement or feature. (MAJOR)
- `Docs` - changes to documentation only.
- `Build` - changes to build process only.
- `Upgrade` - for a dependency upgrade.

An alternative is the commit message style used by Angular and a number of other projects (e.g. `chore(upgrade): react to 0.14`)

### CODE REVIEW

File: scripts/commit-msg [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    if (!TAG_REGEX.test(commitMsgs[0])) {
+      echo([" - Commit summary must start with one of:",
+        "    'Fix:'",
+        "    'Update:'",
+        "    'Breaking:'",
+        "    'Docs:'",
+        "    'Build:'",
+        "    'New:'",
+        "    'Upgrade:'",
+        "   Please refer to the contribution guidelines for more details."].join("\n"));
```
Comment (omgaz) [inline scripts/commit-msg]: Please add contribution guidelines :)

File: scripts/commit-msg [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+function checkGitCommit() {
+  var commitMsgs, failed;
+
+  if (hasBranch("master")) {
+    commitMsgs = splitCommandResultToLines(execSilent("git log HEAD --not master --format=format:%s --no-merges"));
+  } else {
+    commitMsgs = [execSilent("git log -1 --format=format:%s --no-merges")];
+  }
+
+  echo("Validating commit message...");
```
Comment (omgaz) [inline scripts/commit-msg]: …

File: scripts/commit-msg (hunk 1 above)
Comment (jamesramsay) [inline scripts/commit-msg]: Will do, once we agree on actual style :wink:

File: scripts/commit-msg [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        branchName = branchesRaw[i].replace(/^\*(.*)/, "$1").trim();
+        branches.push(branchName);
+    }
+    return branches;
+}
+
+function checkGitCommit() {
+  var commitMsgs, failed;
+
+  if (hasBranch("master")) {
```
Comment (omgaz) [inline scripts/commit-msg]: Why do we care about the existence of master?

**Comment by runk** [comment]
LGTM

**Comment by omgaz** [comment]
Should this not be a pre-commit hook? How does post-test get run?

**Comment by omgaz** [comment]
LGTM Merging :ramen:

**Pull request merged**